### PR TITLE
Backport [BEAM-745]: Add Display Data 'path' metadata

### PR DESCRIPTION
**Pull request by swegner**

### CODE REVIEW

**Comment by swegner** [comment]
R: @lukecwik 
CC: @bjchambers

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/dataflow/DataflowUnboundedReadFromBoundedSource.java [hunk 1]
```diff
@@ -188,13 +188,6 @@ public Reader createReader(PipelineOptions options, Checkpoint<T> checkpoint) {
       return new CheckpointCoder<>(boundedSource.getDefaultOutputCoder());
     }
 
-    @Override
-    public void populateDisplayData(DisplayData.Builder builder) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/dataflow/DataflowUnboundedReadFromBoundedSource.java]: This also differs from Beam/PR/1088

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/Combine.java [hunk 2]
```diff
@@ -1418,10 +1418,12 @@ public void processElement(DoFn<Void, OutputT>.ProcessContext c) {
 
   private static void populateDisplayData(
       DisplayData.Builder builder, HasDisplayData fn,
-      DisplayData.Item<? extends Class<?>> fnDisplayItem) {
-    builder
-        .include(fn)
-        .add(fnDisplayItem);
+      DisplayData.ItemSpec<? extends Class<?>> fnDisplayItem) {
+    if (fn != null) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/Combine.java]: Why can this be null here yet always not null in Beam?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/MapElements.java [hunk 3]
```diff
@@ -111,7 +111,7 @@ public void processElement(ProcessContext c) {
 
       @Override
       public void populateDisplayData(DisplayData.Builder builder) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/MapElements.java]: This class differs from Beam/PR/1088

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 4]
```diff
@@ -100,11 +109,11 @@ public static Type inferType(@Nullable Object value) {
   }
 
   @JsonValue
-  public Collection<Item<?>> items() {
+  public Collection<Item> items() {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 5]
```diff
@@ -100,11 +109,11 @@ public static Type inferType(@Nullable Object value) {
...
     return entries.values();
   }
 
-  public Map<Identifier, Item<?>> asMap() {
+  public Map<Identifier, Item> asMap() {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 6]
```diff
@@ -150,117 +159,111 @@ private static String namespaceOf(Class<?> clazz) {
...
+     * <p>Using {@code include(path, subcomponent)} will associate each of the registered items with
+     * the namespace of the {@code subcomponent} being registered, with the specified path element
+     * relative to the current path. To register display data in the current path and namespace,
+     * such as from a base class implementation, use
      * {@code subcomponent.populateDisplayData(builder)} instead.
      *
      * @see HasDisplayData#populateDisplayData(DisplayData.Builder)
      */
-    Builder include(HasDisplayData subComponent);
+    Builder include(String path, HasDisplayData subComponent);
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 7]
```diff
@@ -150,117 +159,111 @@ private static String namespaceOf(Class<?> clazz) {
...
+     * </code></pre>
      */
-    Builder include(HasDisplayData subComponent, String namespace);
+    Builder delegate(HasDisplayData component);
 
     /**
      * Register the given display item.
      */
-    Builder add(Item<?> item);
+    Builder add(ItemSpec<?> item);
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 8]
```diff
@@ -150,117 +159,111 @@ private static String namespaceOf(Class<?> clazz) {
...
      * Register the given display item if the value is not null.
      */
-    Builder addIfNotNull(Item<?> item);
+    Builder addIfNotNull(ItemSpec<?> item);
 
     /**
      * Register the given display item if the value is different than the specified default.
      */
-    <T> Builder addIfNotDefault(Item<T> item, @Nullable T defaultValue);
+    <T> Builder addIfNotDefault(ItemSpec<T> item, @Nullable T defaultValue);
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 9]
```diff
@@ -150,117 +159,111 @@ private static String namespaceOf(Class<?> clazz) {
...
      * Register the given display item.
      */
-    Builder add(Item<?> item);
+    Builder add(ItemSpec<?> item);
 
     /**
      * Register the given display item if the value is not null.
      */
-    Builder addIfNotNull(Item<?> item);
+    Builder addIfNotNull(ItemSpec<?> item);
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 10]
```diff
@@ -150,117 +159,111 @@ private static String namespaceOf(Class<?> clazz) {
...
      *
-     * @see #include(HasDisplayData)
+     * <pre><code>{@literal @Override}
+     * public void populateDisplayData(DisplayData.Builder builder) {
+     *   builder.delegate(wrapped);
+     * }
+     * </code></pre>
      */
-    Builder include(HasDisplayData subComponent, String namespace);
+    Builder delegate(HasDisplayData component);
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 11]
```diff
@@ -150,117 +159,111 @@ private static String namespaceOf(Class<?> clazz) {
...
-  public static final class Item<T> implements Serializable {
-    @Nullable private final String namespace;
-    private final String key;
-    private final Type type;
-    @Nullable private final Object value;
-    @Nullable private final Object shortValue;
-    @Nullable private final String label;
-    @Nullable private final String linkUrl;
+  @AutoValue
+  public abstract static class Item {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 12]
```diff
@@ -361,118 +437,166 @@ public String getLinkUrl() {
...
-    private final String key;
+  @AutoValue
+  public abstract static class Identifier {
+    public abstract Path getPath();
+    public abstract Class<?> getNamespace();
+    public abstract String getKey();
 
-    public static Identifier of(Class<?> namespace, String key) {
-      return of(namespaceOf(namespace), key);
+    public static Identifier of(Path path, Class<?> namespace, String key) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 13]
```diff
@@ -607,111 +731,127 @@ private FormattedItemValue(Object longValue, Object shortValue) {
...
-      InternalBuilder builder = new InternalBuilder();
-      builder.include(instance);
-      return builder;
+      this.visitedComponents = Sets.newIdentityHashSet();
+      this.visitedPathMap = Maps.newHashMap();
     }
 
     @Override
-    public Builder include(HasDisplayData subComponent) {
+    public Builder include(String path, HasDisplayData subComponent) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 14]
```diff
@@ -607,111 +731,127 @@ private FormattedItemValue(Object longValue, Object shortValue) {
...
-    private class PopulateDisplayDataException extends RuntimeException {
+    private static class PopulateDisplayDataException extends RuntimeException {
       PopulateDisplayDataException(String message, Throwable cause) {
         super(message, cause);
       }
     }
 
     @Override
-    public Builder add(Item<?> item) {
+    public Builder add(ItemSpec<?> item) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 15]
```diff
@@ -607,111 +731,127 @@ private FormattedItemValue(Object longValue, Object shortValue) {
...
       checkNotNull(item, "Input display item cannot be null");
       return addItemIf(true, item);
     }
 
     @Override
-    public Builder addIfNotNull(Item<?> item) {
+    public Builder addIfNotNull(ItemSpec<?> item) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 16]
```diff
@@ -607,111 +731,127 @@ private FormattedItemValue(Object longValue, Object shortValue) {
...
       checkNotNull(item, "Input display item cannot be null");
       return addItemIf(item.getValue() != null, item);
     }
 
     @Override
-    public <T> Builder addIfNotDefault(Item<T> item, @Nullable T defaultValue) {
+    public <T> Builder addIfNotDefault(ItemSpec<T> item, @Nullable T defaultValue) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 17]
```diff
@@ -725,63 +865,63 @@ private DisplayData build() {
   /**
    * Create a display item for the specified key and string value.
    */
-  public static Item<String> item(String key, @Nullable String value) {
+  public static ItemSpec<String> item(String key, @Nullable String value) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 18]
```diff
@@ -725,63 +865,63 @@ private DisplayData build() {
...
     return item(key, Type.STRING, value);
   }
 
   /**
    * Create a display item for the specified key and integer value.
    */
-  public static Item<Integer> item(String key, @Nullable Integer value) {
+  public static ItemSpec<Integer> item(String key, @Nullable Integer value) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 19]
```diff
@@ -725,63 +865,63 @@ private DisplayData build() {
...
     return item(key, Type.INTEGER, value);
   }
 
   /**
    * Create a display item for the specified key and integer value.
    */
-  public static Item<Long> item(String key, @Nullable Long value) {
+  public static ItemSpec<Long> item(String key, @Nullable Long value) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 20]
```diff
@@ -725,63 +865,63 @@ private DisplayData build() {
...
     return item(key, Type.INTEGER, value);
   }
 
   /**
    * Create a display item for the specified key and floating point value.
    */
-  public static Item<Float> item(String key, @Nullable Float value) {
+  public static ItemSpec<Float> item(String key, @Nullable Float value) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 21]
```diff
@@ -725,63 +865,63 @@ private DisplayData build() {
...
     return item(key, Type.FLOAT, value);
   }
 
   /**
    * Create a display item for the specified key and floating point value.
    */
-  public static Item<Double> item(String key, @Nullable Double value) {
+  public static ItemSpec<Double> item(String key, @Nullable Double value) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 22]
```diff
@@ -725,63 +865,63 @@ private DisplayData build() {
...
     return item(key, Type.FLOAT, value);
   }
 
   /**
    * Create a display item for the specified key and boolean value.
    */
-  public static Item<Boolean> item(String key, @Nullable Boolean value) {
+  public static ItemSpec<Boolean> item(String key, @Nullable Boolean value) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 23]
```diff
@@ -725,63 +865,63 @@ private DisplayData build() {
...
     return item(key, Type.BOOLEAN, value);
   }
 
   /**
    * Create a display item for the specified key and timestamp value.
    */
-  public static Item<Instant> item(String key, @Nullable Instant value) {
+  public static ItemSpec<Instant> item(String key, @Nullable Instant value) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 24]
```diff
@@ -725,63 +865,63 @@ private DisplayData build() {
...
     return item(key, Type.TIMESTAMP, value);
   }
 
   /**
    * Create a display item for the specified key and duration value.
    */
-  public static Item<Duration> item(String key, @Nullable Duration value) {
+  public static ItemSpec<Duration> item(String key, @Nullable Duration value) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 25]
```diff
@@ -725,63 +865,63 @@ private DisplayData build() {
...
     return item(key, Type.DURATION, value);
   }
 
   /**
    * Create a display item for the specified key and class value.
    */
-  public static <T> Item<Class<T>> item(String key, @Nullable Class<T> value) {
+  public static <T> ItemSpec<Class<T>> item(String key, @Nullable Class<T> value) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java [hunk 26]
```diff
@@ -795,10 +935,10 @@ private DisplayData build() {
    *
    *  @see Type#inferType(Object)
    */
-  public static <T> Item<T> item(String key, Type type, @Nullable T value) {
+  public static <T> ItemSpec<T> item(String key, Type type, @Nullable T value) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java]: Backwards incompatible change.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java [hunk 27]
```diff
@@ -161,42 +163,31 @@ protected boolean matchesSafely(DisplayData data, Description mismatchDescriptio
...
-    return includesDisplayDataFrom(subComponent, namespace.getName());
-  }
-
   /**
    * Create a matcher that matches if the examined {@link DisplayData} contains all display data
    * registered from the specified subcomponent and namespace.
    */
-  public static Matcher<DisplayData> includesDisplayDataFrom(
-      final HasDisplayData subComponent, final String namespace) {
+  public static Matcher<DisplayData> includesDisplayDataFor(
```
Comment (lukecwik) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java]: Backwards incompatible change for tests.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java [hunk 28]
```diff
@@ -255,59 +246,59 @@ void missing(Item<?> subsetItem) {
    * Creates a matcher that matches if the examined {@link DisplayData.Item} contains a key
    * with the specified value.
    */
-  public static Matcher<DisplayData.Item<?>> hasKey(String key) {
-    return hasKey(Matchers.is(key));
+  public static Matcher<DisplayData.Item> hasKey(String key) {
```
Comment (lukecwik) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java]: Backwards incompatible change for tests.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java [hunk 29]
```diff
@@ -255,59 +246,59 @@ void missing(Item<?> subsetItem) {
...
+    return hasKey(is(key));
   }
 
   /**
    * Creates a matcher that matches if the examined {@link DisplayData.Item} contains a key
    * matching the specified key matcher.
    */
-  public static Matcher<DisplayData.Item<?>> hasKey(Matcher<String> keyMatcher) {
-    return new FeatureMatcher<DisplayData.Item<?>, String>(keyMatcher, "with key", "key") {
+  public static Matcher<DisplayData.Item> hasKey(Matcher<String> keyMatcher) {
```
Comment (lukecwik) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java]: Backwards incompatible change for tests.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java [hunk 30]
```diff
@@ -255,59 +246,59 @@ void missing(Item<?> subsetItem) {
...
+
   /**
-   * Creates a matcher that matches if the examined {@link DisplayData.Item} contains the
-   * specified namespace.
+   * Creates a matcher that matches if the examined {@link DisplayData.Item} contains a path
+   * matching the specified namespace.
    */
-  public static Matcher<DisplayData.Item<?>> hasNamespace(Class<?> namespace) {
-    return hasNamespace(Matchers.<Class<?>>is(namespace));
+  public static Matcher<DisplayData.Item> hasPath(String... paths) {
```
Comment (lukecwik) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java]: Backwards incompatible change for tests.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java [hunk 31]
```diff
@@ -255,59 +246,59 @@ void missing(Item<?> subsetItem) {
...
    */
-  public static Matcher<DisplayData.Item<?>> hasNamespace(String namespace) {
-    return new FeatureMatcher<DisplayData.Item<?>, String>(
-        Matchers.is(namespace), "display item with namespace", "namespace") {
-      @Override
-      protected String featureValueOf(Item<?> actual) {
-        return actual.getNamespace();
-      }
-    };
+  public static Matcher<DisplayData.Item> hasNamespace(Class<?> namespace) {
```
Comment (lukecwik) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java]: Backwards incompatible change for tests.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java [hunk 32]
```diff
@@ -255,59 +246,59 @@ void missing(Item<?> subsetItem) {
...
   }
 
   /**
    * Creates a matcher that matches if the examined {@link DisplayData.Item} contains a namespace
    * matching the specified namespace matcher.
    */
-  public static Matcher<DisplayData.Item<?>> hasNamespace(Matcher<Class<?>> namespaceMatcher) {
-    return new FeatureMatcher<DisplayData.Item<?>, Class<?>>(
-        namespaceMatcher, "display item with namespace", "namespace") {
+  public static Matcher<DisplayData.Item> hasNamespace(Matcher<Class<?>> namespaceMatcher) {
```
Comment (lukecwik) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java]: Backwards incompatible change for tests.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java [hunk 33]
```diff
@@ -316,19 +307,19 @@ protected String featureValueOf(Item<?> actual) {
    * Creates a matcher that matches if the examined {@link DisplayData.Item} matches the
    * specified type.
    */
-  public static Matcher<DisplayData.Item<?>> hasType(DisplayData.Type type) {
-    return hasType(Matchers.is(type));
+  public static Matcher<DisplayData.Item> hasType(DisplayData.Type type) {
```
Comment (lukecwik) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java]: Backwards incompatible change for tests.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java [hunk 34]
```diff
@@ -316,19 +307,19 @@ protected String featureValueOf(Item<?> actual) {
...
+    return hasType(is(type));
   }
 
   /**
    * Creates a matcher that matches if the examined {@link DisplayData.Item} has a type
    * matching the specified type matcher.
    */
-  public static Matcher<DisplayData.Item<?>> hasType(Matcher<DisplayData.Type> typeMatcher) {
-    return new FeatureMatcher<DisplayData.Item<?>, DisplayData.Type>(
+  public static Matcher<DisplayData.Item> hasType(Matcher<DisplayData.Type> typeMatcher) {
```
Comment (lukecwik) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java]: Backwards incompatible change for tests.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java [hunk 35]
```diff
@@ -339,23 +330,45 @@ protected String featureValueOf(Item<?> actual) {
    * value.
    */
 
-  public static Matcher<DisplayData.Item<?>> hasValue(Object value) {
-    return hasValue(Matchers.is(value));
+  public static Matcher<DisplayData.Item> hasValue(Object value) {
```
Comment (lukecwik) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java]: Backwards incompatible change for tests.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java [hunk 36]
```diff
@@ -339,23 +330,45 @@ protected String featureValueOf(Item<?> actual) {
...
+    return hasValue(is(value));
   }
 
   /**
    * Creates a matcher that matches if the examined {@link DisplayData.Item} contains a value
    * matching the specified value matcher.
    */
-  public static <T> Matcher<DisplayData.Item<?>> hasValue(Matcher<T> valueMatcher) {
-    return new FeatureMatcher<DisplayData.Item<?>, T>(
+  public static <T> Matcher<DisplayData.Item> hasValue(Matcher<T> valueMatcher) {
```
Comment (lukecwik) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayDataMatchers.java]: Backwards incompatible change for tests.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/bigtable/BigtableIO.java [hunk 37]
```diff
@@ -310,7 +310,7 @@ public void populateDisplayData(DisplayData.Builder builder) {
       super.populateDisplayData(builder);
 
       builder.add(DisplayData.item("tableId", tableId)
-        .withLinkUrl("Table ID"));
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/bigtable/BigtableIO.java]: Differs from Beam/PR/1088

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/datastore/DatastoreV1.java [hunk 38]
```diff
@@ -596,7 +596,7 @@ public void processElement(ProcessContext c) throws Exception {
       }
 
       @Override
-      public void populateDisplayData(Builder builder) {
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/datastore/DatastoreV1.java]: Differs from Beam/PR/1088

**Comment by swegner** [comment]
Closing for now as this introduces backwards incompatibilities

**Pull request closed without merging**